src/Storages/S3Queue/StorageS3Queue.cpp

```diff
@@ -143,11 +143,17 @@ StorageS3Queue::StorageS3Queue(
     StorageInMemoryMetadata storage_metadata;
     if (columns_.empty())
     {
-        auto columns = StorageS3::getTableStructureFromDataImpl(configuration, format_settings, context_);
+        ColumnsDescription columns;
+        if (configuration.format == "auto")
+            std::tie(columns, configuration.format) = StorageS3::getTableStructureAndFormatFromData(configuration, format_settings, context_);
+        else
+            columns = StorageS3::getTableStructureFromData(configuration, format_settings, context_);
         storage_metadata.setColumns(columns);
     }
     else
     {
+        if (configuration.format == "auto")
+            configuration.format = StorageS3::getTableStructureAndFormatFromData(configuration, format_settings, context_).second;
```

Comment:
Is it okay that we still need to do all the parsing even if columns are explicitly specified?

Reply:
But we still don't know the format, so we will anyway need to do some parsing.
Yes, with known structure we can do something better, for example try to use format readers instead of schema inference as we know the structure, and it will find the best format for sure. But I considered this use case with unknown format and known structure as really rare to be used by users and didn't want to implement separate logic for it